{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import time\n",
    "import gzip\n",
    "from concurrent import futures\n",
    "from functools import partial\n",
    "import pandas as pd\n",
    "\n",
    "from CHARMio import parse_pairs, parse_gtf, write_pairs\n",
    "from . import ref\n",
    "\n",
    "def cli(args)->int:\n",
    "    filename, gtf_file, out_name, thread = \\\n",
    "        args.filename[0], args.gtf_filename, args.out_name, args.num_thread\n",
    "    # parsing ref gtf and pairs file\n",
    "    pairs = parse_pairs(filename)\n",
    "    # build in-mem exon index\n",
    "    gtf = parse_gtf(gtf_file)\n",
    "    ref = build_in_memory_index(get_exon(gtf))\n",
    "    # do search\n",
    "    cleaned = clean_splicing(pairs, ref, thread)\n",
    "    write_pairs(cleaned, out_name)\n",
    "def get_exon(gtf:pd.DataFrame) -> pd.DataFrame:\n",
    "    # extract exon-gene_name from gtf table\n",
    "    relevant = gtf.query('feature == \"exon\" & source == \"HAVANA\"') #using HAVANA only\n",
    "    gene_id = relevant[\"group\"].str.extract('gene_id \"(ENSG[0-9]{11}.[0-9])\";') #extract gene name from group\n",
    "    gene_id.columns = [\"gene_id\"] # extract returns dataframe rather than series\n",
    "    # don't mind strand\n",
    "    return pd.concat([relevant.drop([\"group\",\"feature\",\"source\",\"score\",\"strand\",\"frame\"],axis=1),gene_id],axis=1)\n",
    "def build_in_memory_index(exons:pd.DataFrame) -> dict:\n",
    "    # split by chr and using IntervalIndex to enable searching\n",
    "    ref_dict = {key : value for key, value in exons.groupby(\"seqname\")}\n",
    "    # build index by chromosome\n",
    "    for chromosome in ref_dict:\n",
    "        # using start, end attrs as index\n",
    "        by_chr_table = ref_dict[chromosome]\n",
    "        bed_tuple = by_chr_table.set_index(['start','end']).index \n",
    "        bed_interval_index = pd.IntervalIndex.from_tuples(bed_tuple)\n",
    "        by_chr_table.index = bed_interval_index\n",
    "        ref_dict[chromosome] = by_chr_table.drop([\"start\",\"end\",\"seqname\"],axis=1)\n",
    "    sys.stderr.write(\"hires_utils::clean_splicing: index done.\\n\")\n",
    "    return ref_dict\n",
    "def legs_co_gene(contact:pd.Series, chromosome:str, ref_dict:dict)->bool:\n",
    "    # whether two legs of contacts are in same gene's exon\n",
    "    # must be intra contacts\n",
    "    pos1, pos2 = contact[\"pos1\"], contact[\"pos2\"]\n",
    "    result = ref_dict[chromosome][ref_dict[chromosome].index.contains(pos1)]\n",
    "    if len(result[result.index.contains(pos2)]) > 0:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "def search_chromosome(contacts_at_chromosome:tuple, ref:dict)->pd.DataFrame:\n",
    "    # search whole chromosome using F::legs_co_gene\n",
    "    # single chr searching for multi-process calling\n",
    "    chromosome, contacts = contacts_at_chromosome[0], contacts_at_chromosome[1]\n",
    "    hit_index = contacts.apply(legs_co_gene, chromosome=chromosome, ref_dict=ref, axis=1)\n",
    "    return contacts[hit_index]\n",
    "\n",
    "def clean_splicing(pairs:pd.DataFrame, ref:dict, thread:int)->pd.DataFrame:\n",
    "    '''\n",
    "    clean contacts from splicing\n",
    "    '''\n",
    "    intra = pairs.loc[pairs['chr1'] == pairs['chr2']] # only search for intra\n",
    "    working_func = partial(search_chromosome, ref=ref) # pool.map can't take multiple iterable as arguments\n",
    "    input_data = [(chromosome, per_chr_contacts) for chromosome, per_chr_contacts in intra.groupby(\"chr1\")] # pool.map can't take additional keyword argument\n",
    "    sys.stderr.write(\"hires_utils::clean_splicing: input parsed, search in %d thread\\n\" % thread)\n",
    "    # do multi-thread searching\n",
    "    with futures.ProcessPoolExecutor(thread) as pool:\n",
    "        res = pool.map(working_func, input_data)\n",
    "    result = pd.concat(res, axis=0) # contained contacts\n",
    "    cleaned = pairs.drop(result.index, axis=0) # clean contacts\n",
    "    print(\"clean_splicing: %d contacts removed in %s\\n\" %(len(result), pairs.attrs[\"name\"]) )\n",
    "    return cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "hires_utils::clean_splicing: index done.\n"
     ]
    }
   ],
   "source": [
    "infile = \"/shareb/zliu/project/202212/hires_pipe/result/cleaned_pairs/c12/OrgfE951001.pairs.gz\"\n",
    "gtf_file = \"/share/Data/public/ref_genome/mouse_ref/M23/raw_data/annotation.gtf\"\n",
    "outfile = \"/shareb/zliu/project/202212/hires_pipe/result/cleaned_pairs/c123/OrgfE951001.pairs.gz\"\n",
    "thread = 6\n",
    "\n",
    "pairs = parse_pairs(infile)\n",
    "# build in-mem exon index\n",
    "gtf = parse_gtf(gtf_file)\n",
    "ref = build_in_memory_index(get_exon(gtf))\n",
    "# do search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "readID     category\n",
       "chr1       category\n",
       "pos1          int64\n",
       "chr2       category\n",
       "pos2          int64\n",
       "strand1    category\n",
       "strand2    category\n",
       "phase0     category\n",
       "phase1     category\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#show column types of pairs\n",
    "pairs.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readID</th>\n",
       "      <th>chr1</th>\n",
       "      <th>pos1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>pos2</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>phase0</th>\n",
       "      <th>phase1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3001595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>63433201</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3009219</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24592399</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3009236</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4224042</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3010458</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4337937</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3029765</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3066833</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162950</th>\n",
       "      <td>.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169901613</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169907419</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162951</th>\n",
       "      <td>.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169902685</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169904922</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162952</th>\n",
       "      <td>.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169912697</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169916734</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162953</th>\n",
       "      <td>.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169934456</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169956108</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162954</th>\n",
       "      <td>.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169959627</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169964994</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>140807 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       readID  chr1       pos1  chr2       pos2 strand1 strand2 phase0 phase1\n",
       "0           .  chr1    3001595  chr1   63433201       +       +      0      0\n",
       "1           .  chr1    3009219  chr1   24592399       +       +      1      1\n",
       "2           .  chr1    3009236  chr1    4224042       +       +      .      .\n",
       "3           .  chr1    3010458  chr1    4337937       +       +      0      0\n",
       "4           .  chr1    3029765  chr1    3066833       +       +      .      1\n",
       "...       ...   ...        ...   ...        ...     ...     ...    ...    ...\n",
       "162950      .  chrX  169901613  chrX  169907419       +       +      0      0\n",
       "162951      .  chrX  169902685  chrX  169904922       +       +      .      .\n",
       "162952      .  chrX  169912697  chrX  169916734       +       +      .      1\n",
       "162953      .  chrX  169934456  chrX  169956108       +       +      .      .\n",
       "162954      .  chrX  169959627  chrX  169964994       +       +      .      .\n",
       "\n",
       "[140807 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#pairs.query('chr1 == chr2')\n",
    "pairs.loc[pairs['chr1'] == pairs['chr2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "hires_utils::clean_splicing: input parsed, search in 6 thread\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clean_splicing: 80 contacts removed in OrgfE951001\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cleaned = clean_splicing(pairs, ref, thread)\n",
    "write_pairs(cleaned, outfile)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hic2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "644476859734f9a06a44271925db4b71b460672dce71d1af3cb8b5924b1a2d95"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
