{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  gene = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` for the full column specifications.\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  gene = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` for the full column specifications.\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X3 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X4 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X5 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "Joining, by = \"cellname\"\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(ggpubr)\n",
    "library(GGally)\n",
    "library(yaml)\n",
    "library(patchwork)\n",
    "#library(Seurat)\n",
    "#library(ggvenn)\n",
    "\n",
    "#read config\n",
    "config <- read_yaml(file = \"../CHARM//config.yaml\")\n",
    "\n",
    "#Common\n",
    "Raw <- read_table2(\"../stat/raw.fq.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X2 = X2/4*300/1000000000,X1 = strsplit(X1,split = \"/\")[[1]][3],X1 = strsplit(X1,split = \"_\")[[1]][1])\n",
    "DNA <- read_table2(\"../stat/dna.fq.stat\",col_names=F) %>% arrange(X1) %>% rowwise()%>% mutate(X2 = X2/4*300/1000000000,X1 = strsplit(X1,split = \"/\")[[1]][3])\n",
    "RNA <- read_table2(\"../stat/rna.fq.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X2 = X2/4*300/1000000000,X1 = strsplit(X1,split = \"/\")[[1]][3])\n",
    "\n",
    "raw_pairs <- read_table2(\"../stat/raw.pairs.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][3])\n",
    "pairs_dedup <- read_table2(\"../stat/pairs.dedup.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][3])\n",
    "\n",
    "pairs_c1 <- read_table2(\"../stat/pairs.c1.stat\",col_names=F)%>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "pairs_c12 <- read_table2(\"../stat/pairs.c12.stat\",col_names=F)%>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "pairs_c123 <- read_table2(\"../stat/pairs.c123.stat\",col_names=F)%>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "inter_pairs_c123 <- read_table2(file = \"../stat/inter.pairs.c123.stat\",col_names = F) %>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "\n",
    "yperx <- read_table2(\"../stat/yperx.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][2])\n",
    "stat <- Raw %>% left_join(DNA,by=\"X1\") %>% left_join(RNA,by=\"X1\") %>% left_join(yperx,by=\"X1\") %>% left_join(raw_pairs,by=\"X1\") %>% left_join(pairs_dedup,by=\"X1\") %>% \n",
    "    left_join(pairs_c1,by=\"X1\") %>% left_join(pairs_c12,by=\"X1\")%>% left_join(pairs_c123,by=\"X1\") %>% left_join(inter_pairs_c123,by=\"X1\")\n",
    "RNAres <- read_table2(\"../result/RNA_Res/counts.gene.total.format.tsv\")\n",
    "featureStat <- cbind(as.data.frame(colSums(RNAres %>% select(-gene))),colSums(RNAres %>% select(-gene) != 0)) %>% rownames_to_column(\"X1\")\n",
    "RNAresExon <- read_table2(\"../result/RNA_Res/counts.exon.total.format.tsv\")\n",
    "featureStatExon <- cbind(as.data.frame(colSums(RNAresExon %>% select(-gene))),colSums(RNAresExon %>% select(-gene) != 0))  %>% rownames_to_column(\"X1\")\n",
    "\n",
    "if(config$if_RNA_snp_split){\n",
    "    RNAresGenome1 <- read_table2(\"../result/RNA_Res/counts.gene.genome1.tsv\")\n",
    "    RNAresGenome2 <- read_table2(\"../result/RNA_Res/counts.gene.genome2.tsv\")\n",
    "    featureStatGenome1 <- cbind(as.data.frame(colSums(RNAresGenome1 %>% select(-gene))),colSums(RNAresGenome1 %>% select(-gene) != 0)) %>% rownames_to_column(\"X1\")\n",
    "    featureStatGenome2 <- cbind(as.data.frame(colSums(RNAresGenome2 %>% select(-gene))),colSums(RNAresGenome2 %>% select(-gene) != 0)) %>% rownames_to_column(\"X1\")\n",
    "    stat <- stat %>% left_join(featureStat,by=\"X1\") %>% left_join(featureStatExon,by=\"X1\") %>% left_join(featureStatGenome1,by=\"X1\")  %>% left_join(featureStatGenome2,by=\"X1\")\n",
    "    names(stat) <- c(\"cellname\",\"Rawreads\",\"DNAreads\",\"RNAreads\",\"yperx\",\"raw_pairs\",\"pairs_dedup\",\"pairs_clean1\",\"pairs_clean2\",\"pairs_clean3\",\"inter_pairs_clean3\",\n",
    "                     \"UMIs_gene\",\"genes_gene\",\"UMIs_exon\",\"genes_exon\",\"UMIs_gene_genome1\",\"genes_gene_genome1\",\"UMIs_gene_genome2\",\"genes_gene_genome2\")\n",
    "} else{\n",
    "    stat <- stat %>% left_join(featureStat,by=\"X1\") %>% left_join(featureStatExon,by=\"X1\") \n",
    "    names(stat) <- c(\"cellname\",\"Rawreads\",\"DNAreads\",\"RNAreads\",\"yperx\",\"raw_pairs\",\"pairs_dedup\",\"pairs_clean1\",\"pairs_clean2\",\"pairs_clean3\",\"inter_pairs_clean3\",\n",
    "                     \"UMIs_gene\",\"genes_gene\",\"UMIs_exon\",\"genes_exon\")\n",
    "}\n",
    "\n",
    "\n",
    "if(config$if_structure){\n",
    "    rmsd <- read_table2(\"../stat/rmsd.info\",col_names = F) %>% rowwise() %>% mutate(X1=str_split(X1,pattern = fixed(\"/\")),cellname = X1[1], res = str_split(X1[3], fixed(\".\"))[[1]][2],TOP3_RMSD = X5) %>% select(cellname,res,TOP3_RMSD) %>% spread(res,TOP3_RMSD) %>% arrange(cellname)\n",
    "    stat <- stat %>% left_join(rmsd)\n",
    "}\n",
    "if (config$if_cuttag){\n",
    "    reads <- tibble(\"cellname\"=NA)\n",
    "    for(i in 1:length(names(config$split))){\n",
    "        temp <- read_table2(paste0(\"../stat/\",names(config$split)[i],\".read.stat\"),col_names = F) %>% rowwise() %>% mutate(X1 = str_split(X1,pattern =\"/\")[[1]][2],X2 = X2/4*300/1000000000)\n",
    "        names(temp) <- c(\"cellname\",paste0(names(config$split)[i],\"_reads\"))\n",
    "        reads <- reads %>% full_join(temp)\n",
    "    }\n",
    "    reads <- reads %>% filter(!is.na(cellname))\n",
    "\n",
    "    frag <- tibble(\"cellname\"=NA)\n",
    "    for(i in 1:length(names(config$split))){\n",
    "        temp <- read_table2(paste0(\"../stat/\",names(config$split)[i],\".frag.stat\"),col_names = F) %>% rowwise() %>% mutate(X1 = str_extract(X1,pattern = \"(?<=\\\\/)[A-Za-z0-9-]+(?=\\\\.)\"))\n",
    "        names(temp) <- c(\"cellname\",paste0(names(config$split)[i],\"_frag\"))\n",
    "        frag <- frag %>% full_join(temp)\n",
    "    }\n",
    "    frag <- frag %>% filter(!is.na(cellname))\n",
    "    reads <- reads %>% left_join(frag)\n",
    "    stat <- stat %>% full_join(reads)\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot <- stat %>% mutate(RNAreadsRatio = RNAreads/(RNAreads+DNAreads),\n",
    "                        pairsPerRead = raw_pairs/DNAreads/1000000000*300,\n",
    "                        pairsValidRatio = pairs_clean3/raw_pairs,\n",
    "                        interPairsRatio = inter_pairs_clean3/pairs_clean3) \n",
    "#plot[is.na(plot)]<-0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   cellname            Rawreads         DNAreads         RNAreads       \n",
       " Length:192         Min.   :0.3023   Min.   :0.2973   Min.   :0.002575  \n",
       " Class :character   1st Qu.:2.4015   1st Qu.:2.2635   1st Qu.:0.046244  \n",
       " Mode  :character   Median :4.1420   Median :3.8554   Median :0.152186  \n",
       "                    Mean   :4.0641   Mean   :3.8001   Mean   :0.230230  \n",
       "                    3rd Qu.:5.5827   3rd Qu.:5.1283   3rd Qu.:0.336940  \n",
       "                    Max.   :8.5390   Max.   :8.1782   Max.   :1.275965  \n",
       "     yperx            raw_pairs        pairs_dedup      pairs_clean1   \n",
       " Min.   :0.000965   Min.   :  65874   Min.   : 51698   Min.   : 51559  \n",
       " 1st Qu.:0.005006   1st Qu.: 662960   1st Qu.:307964   1st Qu.:305150  \n",
       " Median :0.005871   Median :1322358   Median :373257   Median :368597  \n",
       " Mean   :0.007437   Mean   :1363911   Mean   :375294   Mean   :371832  \n",
       " 3rd Qu.:0.006839   3rd Qu.:1912474   3rd Qu.:448309   3rd Qu.:444313  \n",
       " Max.   :0.095524   Max.   :3351396   Max.   :671798   Max.   :666881  \n",
       "  pairs_clean2     pairs_clean3    inter_pairs_clean3   UMIs_gene    \n",
       " Min.   : 38278   Min.   : 38256   Min.   :  2358     Min.   : 1089  \n",
       " 1st Qu.:249524   1st Qu.:249382   1st Qu.: 38669     1st Qu.:12664  \n",
       " Median :300630   Median :300487   Median : 51970     Median :21246  \n",
       " Mean   :306350   Mean   :306195   Mean   : 58491     Mean   :21753  \n",
       " 3rd Qu.:377661   3rd Qu.:377505   3rd Qu.: 76351     3rd Qu.:28429  \n",
       " Max.   :579838   Max.   :579571   Max.   :169865     Max.   :61488  \n",
       "   genes_gene      UMIs_exon       genes_exon         1m          \n",
       " Min.   :  843   Min.   :  288   Min.   : 231   Min.   : 0.00010  \n",
       " 1st Qu.: 4641   1st Qu.: 3034   1st Qu.:1934   1st Qu.: 0.01543  \n",
       " Median : 6148   Median : 4976   Median :2684   Median : 0.02729  \n",
       " Mean   : 5866   Mean   : 5240   Mean   :2627   Mean   : 0.36276  \n",
       " 3rd Qu.: 7262   3rd Qu.: 7064   3rd Qu.:3385   3rd Qu.: 0.05656  \n",
       " Max.   :10587   Max.   :15314   Max.   :5545   Max.   :11.13833  \n",
       "      200k               20k               50k          RNAreadsRatio     \n",
       " Min.   : 0.01925   Min.   : 0.2499   Min.   : 0.1020   Min.   :0.001055  \n",
       " 1st Qu.: 0.09540   1st Qu.: 0.6558   1st Qu.: 0.3288   1st Qu.:0.017698  \n",
       " Median : 0.15178   Median : 0.8845   Median : 0.4713   Median :0.040404  \n",
       " Mean   : 0.79352   Mean   : 2.4823   Mean   : 1.5695   Mean   :0.049147  \n",
       " 3rd Qu.: 0.24502   3rd Qu.: 1.1802   3rd Qu.: 0.6511   3rd Qu.:0.076898  \n",
       " Max.   :20.56410   Max.   :47.1944   Max.   :33.9624   Max.   :0.177651  \n",
       "  pairsPerRead     pairsValidRatio   interPairsRatio  \n",
       " Min.   :0.02890   Min.   :0.05421   Min.   :0.05622  \n",
       " 1st Qu.:0.08101   1st Qu.:0.16257   1st Qu.:0.15142  \n",
       " Median :0.10644   Median :0.30034   Median :0.17539  \n",
       " Mean   :0.10371   Mean   :0.29585   Mean   :0.18256  \n",
       " 3rd Qu.:0.12873   3rd Qu.:0.41799   3rd Qu.:0.20980  \n",
       " Max.   :0.15601   Max.   :0.58075   Max.   :0.36213  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.matrix.max.cols=50, repr.matrix.max.rows=100)\n",
    "plot %>% summary()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.0.3",
   "language": "R",
   "name": "ir403"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
