{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  gene = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` for the full column specifications.\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  gene = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` for the full column specifications.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(ggpubr)\n",
    "library(GGally)\n",
    "library(yaml)\n",
    "library(patchwork)\n",
    "#library(Seurat)\n",
    "#library(ggvenn)\n",
    "\n",
    "#read config\n",
    "config <- read_yaml(file = \"../HiRES_preprocess_pipeline//config.yaml\")\n",
    "\n",
    "#Common\n",
    "Raw <- read_table2(\"../stat/raw.fq.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X2 = X2/4*300/1000000000,X1 = strsplit(X1,split = \"/\")[[1]][3],X1 = strsplit(X1,split = \"_\")[[1]][1])\n",
    "DNA <- read_table2(\"../stat/dna.fq.stat\",col_names=F) %>% arrange(X1) %>% rowwise()%>% mutate(X2 = X2/4*300/1000000000,X1 = strsplit(X1,split = \"/\")[[1]][3])\n",
    "RNA <- read_table2(\"../stat/rna.fq.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X2 = X2/4*300/1000000000,X1 = strsplit(X1,split = \"/\")[[1]][3])\n",
    "\n",
    "raw_pairs <- read_table2(\"../stat/raw.pairs.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][3])\n",
    "pairs_dedup <- read_table2(\"../stat/pairs.dedup.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][4]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "\n",
    "if(config$if_clean){\n",
    "pairs_c1 <- read_table2(\"../stat/pairs.c1.stat\",col_names=F)%>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "pairs_c12 <- read_table2(\"../stat/pairs.c12.stat\",col_names=F)%>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "pairs_c123 <- read_table2(\"../stat/pairs.c123.stat\",col_names=F)%>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "inter_pairs_c123 <- read_table2(file = \"../stat/inter.pairs.c123.stat\",col_names = F) %>% arrange(X1) %>%rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][5]) %>% mutate(X1 = str_replace(X1,\".pairs.gz\",\"\"))\n",
    "\n",
    "yperx <- read_table2(\"../stat/yperx.stat\",col_names=F) %>% arrange(X1) %>% rowwise() %>% mutate(X1 = strsplit(X1,split = \"/\")[[1]][2])\n",
    "stat <- Raw %>% left_join(DNA,by=\"X1\") %>% left_join(RNA,by=\"X1\") %>% left_join(yperx,by=\"X1\") %>% left_join(raw_pairs,by=\"X1\") %>% left_join(pairs_dedup,by=\"X1\") %>% \n",
    "    left_join(pairs_c1,by=\"X1\") %>% left_join(pairs_c12,by=\"X1\")%>% left_join(pairs_c123,by=\"X1\") %>% left_join(inter_pairs_c123,by=\"X1\")\n",
    "} else{\n",
    "    stat <- Raw %>% left_join(DNA,by=\"X1\") %>% left_join(RNA,by=\"X1\") %>% left_join(yperx,by=\"X1\") %>% left_join(raw_pairs,by=\"X1\") %>% left_join(pairs_dedup,by=\"X1\")\n",
    "}\n",
    "\n",
    "RNAres <- read_table2(\"../result/RNA_Res/counts.gene.total.format.tsv\")\n",
    "featureStat <- cbind(as.data.frame(colSums(RNAres %>% select(-gene))),colSums(RNAres %>% select(-gene) != 0)) %>% rownames_to_column(\"X1\")\n",
    "RNAresExon <- read_table2(\"../result/RNA_Res/counts.exon.total.format.tsv\")\n",
    "featureStatExon <- cbind(as.data.frame(colSums(RNAresExon %>% select(-gene))),colSums(RNAresExon %>% select(-gene) != 0))  %>% rownames_to_column(\"X1\")\n",
    "\n",
    "if(config$if_RNA_snp_split){\n",
    "    RNAresGenome1 <- read_table2(\"../result/RNA_Res/counts.gene.genome1.tsv\")\n",
    "    RNAresGenome2 <- read_table2(\"../result/RNA_Res/counts.gene.genome2.tsv\")\n",
    "    featureStatGenome1 <- cbind(as.data.frame(colSums(RNAresGenome1 %>% select(-gene))),colSums(RNAresGenome1 %>% select(-gene) != 0)) %>% rownames_to_column(\"X1\")\n",
    "    featureStatGenome2 <- cbind(as.data.frame(colSums(RNAresGenome2 %>% select(-gene))),colSums(RNAresGenome2 %>% select(-gene) != 0)) %>% rownames_to_column(\"X1\")\n",
    "    stat <- stat %>% left_join(featureStat,by=\"X1\") %>% left_join(featureStatExon,by=\"X1\") %>% left_join(featureStatGenome1,by=\"X1\")  %>% left_join(featureStatGenome2,by=\"X1\")\n",
    "    names(stat) <- c(\"cellname\",\"Rawreads\",\"DNAreads\",\"RNAreads\",\"yperx\",\"raw_pairs\",\"pairs_dedup\",\"pairs_clean1\",\"pairs_clean2\",\"pairs_clean3\",\"inter_pairs_clean3\",\n",
    "                     \"UMIs_gene\",\"genes_gene\",\"UMIs_exon\",\"genes_exon\",\"UMIs_gene_genome1\",\"genes_gene_genome1\",\"UMIs_gene_genome2\",\"genes_gene_genome2\")\n",
    "} else if (config$if_clean){\n",
    "    stat <- stat %>% left_join(featureStat,by=\"X1\") %>% left_join(featureStatExon,by=\"X1\") \n",
    "    names(stat) <- c(\"cellname\",\"Rawreads\",\"DNAreads\",\"RNAreads\",\"yperx\",\"raw_pairs\",\"pairs_dedup\",\"pairs_clean1\",\"pairs_clean2\",\"pairs_clean3\",\"inter_pairs_clean3\",\n",
    "                     \"UMIs_gene\",\"genes_gene\",\"UMIs_exon\",\"genes_exon\")\n",
    "} else {\n",
    "    stat <- stat %>% left_join(featureStat,by=\"X1\") %>% left_join(featureStatExon,by=\"X1\")\n",
    "    names(stat) <- c(\"cellname\",\"Rawreads\",\"DNAreads\",\"RNAreads\",\"yperx\",\"raw_pairs\",\"pairs_dedup\",\"UMIs_gene\",\"genes_gene\",\"UMIs_exon\",\"genes_exon\")\n",
    "}\n",
    "\n",
    "\n",
    "if(config$if_structure){\n",
    "    rmsd <- read_table2(\"../stat/rmsd.info\",col_names = F) %>% rowwise() %>% mutate(X1=str_split(X1,pattern = fixed(\"/\")),cellname = X1[1], res = str_split(X1[3], fixed(\".\"))[[1]][2],TOP3_RMSD = X5) %>% select(cellname,res,TOP3_RMSD) %>% spread(res,TOP3_RMSD) %>% arrange(cellname)\n",
    "    stat <- stat %>% left_join(rmsd)\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 32 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cellname</th><th scope=col>Rawreads</th><th scope=col>DNAreads</th><th scope=col>RNAreads</th><th scope=col>yperx</th><th scope=col>raw_pairs</th><th scope=col>pairs_dedup</th><th scope=col>UMIs_gene</th><th scope=col>genes_gene</th><th scope=col>UMIs_exon</th><th scope=col>genes_exon</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>16nA1</td><td>0.3989055</td><td>0.3957042</td><td>0.0018330</td><td>-1</td><td>234464</td><td>123576</td><td>370</td><td>350</td><td>356</td><td>333</td></tr>\n",
       "\t<tr><td>16nA2</td><td>0.1169457</td><td>0.1164594</td><td>0.0000639</td><td>-1</td><td>  2454</td><td>   254</td><td>  1</td><td>  1</td><td>  1</td><td>  1</td></tr>\n",
       "\t<tr><td>16nA3</td><td>0.7003767</td><td>0.6932880</td><td>0.0039033</td><td>-1</td><td>449639</td><td>184320</td><td>512</td><td>432</td><td>482</td><td>400</td></tr>\n",
       "\t<tr><td>16nA4</td><td>0.6771048</td><td>0.6696726</td><td>0.0039687</td><td>-1</td><td>422654</td><td>181092</td><td>522</td><td>444</td><td>488</td><td>419</td></tr>\n",
       "\t<tr><td>16nA5</td><td>0.1011897</td><td>0.1009551</td><td>0.0000015</td><td>-1</td><td>  2401</td><td>   222</td><td> NA</td><td> NA</td><td> NA</td><td> NA</td></tr>\n",
       "\t<tr><td>16nA6</td><td>0.5432619</td><td>0.5375346</td><td>0.0031887</td><td>-1</td><td>354650</td><td>144051</td><td>321</td><td>266</td><td>311</td><td>257</td></tr>\n",
       "\t<tr><td>16nA7</td><td>0.5378295</td><td>0.5319861</td><td>0.0030447</td><td>-1</td><td>348093</td><td>164068</td><td>399</td><td>349</td><td>384</td><td>335</td></tr>\n",
       "\t<tr><td>16nA8</td><td>0.0892878</td><td>0.0891042</td><td>0.0000072</td><td>-1</td><td>  2586</td><td>   212</td><td>  2</td><td>  1</td><td>  2</td><td>  1</td></tr>\n",
       "\t<tr><td>2nA1 </td><td>0.1032909</td><td>0.1029816</td><td>0.0000183</td><td>-1</td><td>  1985</td><td>   188</td><td> NA</td><td> NA</td><td> NA</td><td> NA</td></tr>\n",
       "\t<tr><td>2nA2 </td><td>0.1311063</td><td>0.1303647</td><td>0.0004314</td><td>-1</td><td> 21142</td><td>  8502</td><td> 73</td><td> 70</td><td> 69</td><td> 64</td></tr>\n",
       "\t<tr><td>2nA3 </td><td>0.1577292</td><td>0.1568262</td><td>0.0000006</td><td>-1</td><td>  2237</td><td>   199</td><td> NA</td><td> NA</td><td> NA</td><td> NA</td></tr>\n",
       "\t<tr><td>2nA4 </td><td>0.1071246</td><td>0.1067223</td><td>0.0000705</td><td>-1</td><td>  2201</td><td>   253</td><td>  1</td><td>  1</td><td>  1</td><td>  1</td></tr>\n",
       "\t<tr><td>2nA5 </td><td>0.1969200</td><td>0.1944762</td><td>0.0015129</td><td>-1</td><td> 74984</td><td> 23197</td><td>146</td><td>113</td><td>135</td><td>103</td></tr>\n",
       "\t<tr><td>2nA6 </td><td>0.1436553</td><td>0.1433478</td><td>0.0000891</td><td>-1</td><td>  3448</td><td>   283</td><td> NA</td><td> NA</td><td> NA</td><td> NA</td></tr>\n",
       "\t<tr><td>2nA7 </td><td>0.0684474</td><td>0.0683364</td><td>0.0000024</td><td>-1</td><td>  1693</td><td>   159</td><td>  3</td><td>  2</td><td>  3</td><td>  2</td></tr>\n",
       "\t<tr><td>2nA8 </td><td>0.1585038</td><td>0.1540401</td><td>0.0033555</td><td>-1</td><td> 51342</td><td> 20866</td><td>381</td><td>321</td><td>356</td><td>298</td></tr>\n",
       "\t<tr><td>4nA1 </td><td>0.0943545</td><td>0.0941154</td><td>0.0000177</td><td>-1</td><td>  2362</td><td>   383</td><td>  2</td><td>  2</td><td>  2</td><td>  2</td></tr>\n",
       "\t<tr><td>4nA2 </td><td>0.1304181</td><td>0.1301577</td><td>0.0000108</td><td>-1</td><td>  2352</td><td>   231</td><td>  2</td><td>  2</td><td>  2</td><td>  2</td></tr>\n",
       "\t<tr><td>4nA3 </td><td>0.2298732</td><td>0.2273217</td><td>0.0015957</td><td>-1</td><td> 98589</td><td> 31109</td><td>142</td><td>111</td><td>130</td><td>102</td></tr>\n",
       "\t<tr><td>4nA4 </td><td>0.4959204</td><td>0.4905402</td><td>0.0030840</td><td>-1</td><td>224717</td><td> 69245</td><td>260</td><td>221</td><td>247</td><td>209</td></tr>\n",
       "\t<tr><td>4nA5 </td><td>0.1029048</td><td>0.1026396</td><td>0.0000255</td><td>-1</td><td>  2464</td><td>   234</td><td>  7</td><td>  4</td><td>  7</td><td>  4</td></tr>\n",
       "\t<tr><td>4nA6 </td><td>0.1025673</td><td>0.1023861</td><td>0.0000012</td><td>-1</td><td>  2742</td><td>   259</td><td>  1</td><td>  1</td><td>  1</td><td>  1</td></tr>\n",
       "\t<tr><td>4nA7 </td><td>0.3244281</td><td>0.3240933</td><td>0.0000024</td><td>-1</td><td>  3088</td><td>   213</td><td>  1</td><td>  1</td><td>  1</td><td>  1</td></tr>\n",
       "\t<tr><td>4nA8 </td><td>0.3660981</td><td>0.3618714</td><td>0.0025242</td><td>-1</td><td>173405</td><td> 56069</td><td>199</td><td>177</td><td>186</td><td>164</td></tr>\n",
       "\t<tr><td>8nA1 </td><td>0.1311906</td><td>0.1308564</td><td>0.0000198</td><td>-1</td><td>  2605</td><td>   250</td><td>  1</td><td>  1</td><td>  1</td><td>  1</td></tr>\n",
       "\t<tr><td>8nA2 </td><td>0.1035069</td><td>0.1031595</td><td>0.0000174</td><td>-1</td><td>  2281</td><td>   234</td><td>  1</td><td>  1</td><td>  1</td><td>  1</td></tr>\n",
       "\t<tr><td>8nA3 </td><td>0.0798825</td><td>0.0797247</td><td>0.0000006</td><td>-1</td><td>  2500</td><td>   286</td><td> NA</td><td> NA</td><td> NA</td><td> NA</td></tr>\n",
       "\t<tr><td>8nA4 </td><td>0.3374520</td><td>0.3343263</td><td>0.0016212</td><td>-1</td><td>185664</td><td> 86932</td><td>248</td><td>221</td><td>239</td><td>212</td></tr>\n",
       "\t<tr><td>8nA5 </td><td>0.3425886</td><td>0.3370527</td><td>0.0036342</td><td>-1</td><td>199713</td><td> 66713</td><td>330</td><td>269</td><td>315</td><td>253</td></tr>\n",
       "\t<tr><td>8nA6 </td><td>0.0903384</td><td>0.0901446</td><td>0.0000009</td><td>-1</td><td>  2789</td><td>   263</td><td> NA</td><td> NA</td><td> NA</td><td> NA</td></tr>\n",
       "\t<tr><td>8nA7 </td><td>0.5560356</td><td>0.5496432</td><td>0.0034056</td><td>-1</td><td>352787</td><td>129172</td><td>356</td><td>293</td><td>340</td><td>281</td></tr>\n",
       "\t<tr><td>8nA8 </td><td>0.4931508</td><td>0.4875861</td><td>0.0030840</td><td>-1</td><td>287061</td><td>108574</td><td>325</td><td>276</td><td>304</td><td>259</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 32 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " cellname & Rawreads & DNAreads & RNAreads & yperx & raw\\_pairs & pairs\\_dedup & UMIs\\_gene & genes\\_gene & UMIs\\_exon & genes\\_exon\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 16nA1 & 0.3989055 & 0.3957042 & 0.0018330 & -1 & 234464 & 123576 & 370 & 350 & 356 & 333\\\\\n",
       "\t 16nA2 & 0.1169457 & 0.1164594 & 0.0000639 & -1 &   2454 &    254 &   1 &   1 &   1 &   1\\\\\n",
       "\t 16nA3 & 0.7003767 & 0.6932880 & 0.0039033 & -1 & 449639 & 184320 & 512 & 432 & 482 & 400\\\\\n",
       "\t 16nA4 & 0.6771048 & 0.6696726 & 0.0039687 & -1 & 422654 & 181092 & 522 & 444 & 488 & 419\\\\\n",
       "\t 16nA5 & 0.1011897 & 0.1009551 & 0.0000015 & -1 &   2401 &    222 &  NA &  NA &  NA &  NA\\\\\n",
       "\t 16nA6 & 0.5432619 & 0.5375346 & 0.0031887 & -1 & 354650 & 144051 & 321 & 266 & 311 & 257\\\\\n",
       "\t 16nA7 & 0.5378295 & 0.5319861 & 0.0030447 & -1 & 348093 & 164068 & 399 & 349 & 384 & 335\\\\\n",
       "\t 16nA8 & 0.0892878 & 0.0891042 & 0.0000072 & -1 &   2586 &    212 &   2 &   1 &   2 &   1\\\\\n",
       "\t 2nA1  & 0.1032909 & 0.1029816 & 0.0000183 & -1 &   1985 &    188 &  NA &  NA &  NA &  NA\\\\\n",
       "\t 2nA2  & 0.1311063 & 0.1303647 & 0.0004314 & -1 &  21142 &   8502 &  73 &  70 &  69 &  64\\\\\n",
       "\t 2nA3  & 0.1577292 & 0.1568262 & 0.0000006 & -1 &   2237 &    199 &  NA &  NA &  NA &  NA\\\\\n",
       "\t 2nA4  & 0.1071246 & 0.1067223 & 0.0000705 & -1 &   2201 &    253 &   1 &   1 &   1 &   1\\\\\n",
       "\t 2nA5  & 0.1969200 & 0.1944762 & 0.0015129 & -1 &  74984 &  23197 & 146 & 113 & 135 & 103\\\\\n",
       "\t 2nA6  & 0.1436553 & 0.1433478 & 0.0000891 & -1 &   3448 &    283 &  NA &  NA &  NA &  NA\\\\\n",
       "\t 2nA7  & 0.0684474 & 0.0683364 & 0.0000024 & -1 &   1693 &    159 &   3 &   2 &   3 &   2\\\\\n",
       "\t 2nA8  & 0.1585038 & 0.1540401 & 0.0033555 & -1 &  51342 &  20866 & 381 & 321 & 356 & 298\\\\\n",
       "\t 4nA1  & 0.0943545 & 0.0941154 & 0.0000177 & -1 &   2362 &    383 &   2 &   2 &   2 &   2\\\\\n",
       "\t 4nA2  & 0.1304181 & 0.1301577 & 0.0000108 & -1 &   2352 &    231 &   2 &   2 &   2 &   2\\\\\n",
       "\t 4nA3  & 0.2298732 & 0.2273217 & 0.0015957 & -1 &  98589 &  31109 & 142 & 111 & 130 & 102\\\\\n",
       "\t 4nA4  & 0.4959204 & 0.4905402 & 0.0030840 & -1 & 224717 &  69245 & 260 & 221 & 247 & 209\\\\\n",
       "\t 4nA5  & 0.1029048 & 0.1026396 & 0.0000255 & -1 &   2464 &    234 &   7 &   4 &   7 &   4\\\\\n",
       "\t 4nA6  & 0.1025673 & 0.1023861 & 0.0000012 & -1 &   2742 &    259 &   1 &   1 &   1 &   1\\\\\n",
       "\t 4nA7  & 0.3244281 & 0.3240933 & 0.0000024 & -1 &   3088 &    213 &   1 &   1 &   1 &   1\\\\\n",
       "\t 4nA8  & 0.3660981 & 0.3618714 & 0.0025242 & -1 & 173405 &  56069 & 199 & 177 & 186 & 164\\\\\n",
       "\t 8nA1  & 0.1311906 & 0.1308564 & 0.0000198 & -1 &   2605 &    250 &   1 &   1 &   1 &   1\\\\\n",
       "\t 8nA2  & 0.1035069 & 0.1031595 & 0.0000174 & -1 &   2281 &    234 &   1 &   1 &   1 &   1\\\\\n",
       "\t 8nA3  & 0.0798825 & 0.0797247 & 0.0000006 & -1 &   2500 &    286 &  NA &  NA &  NA &  NA\\\\\n",
       "\t 8nA4  & 0.3374520 & 0.3343263 & 0.0016212 & -1 & 185664 &  86932 & 248 & 221 & 239 & 212\\\\\n",
       "\t 8nA5  & 0.3425886 & 0.3370527 & 0.0036342 & -1 & 199713 &  66713 & 330 & 269 & 315 & 253\\\\\n",
       "\t 8nA6  & 0.0903384 & 0.0901446 & 0.0000009 & -1 &   2789 &    263 &  NA &  NA &  NA &  NA\\\\\n",
       "\t 8nA7  & 0.5560356 & 0.5496432 & 0.0034056 & -1 & 352787 & 129172 & 356 & 293 & 340 & 281\\\\\n",
       "\t 8nA8  & 0.4931508 & 0.4875861 & 0.0030840 & -1 & 287061 & 108574 & 325 & 276 & 304 & 259\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 32 × 11\n",
       "\n",
       "| cellname &lt;chr&gt; | Rawreads &lt;dbl&gt; | DNAreads &lt;dbl&gt; | RNAreads &lt;dbl&gt; | yperx &lt;dbl&gt; | raw_pairs &lt;dbl&gt; | pairs_dedup &lt;dbl&gt; | UMIs_gene &lt;dbl&gt; | genes_gene &lt;dbl&gt; | UMIs_exon &lt;dbl&gt; | genes_exon &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 16nA1 | 0.3989055 | 0.3957042 | 0.0018330 | -1 | 234464 | 123576 | 370 | 350 | 356 | 333 |\n",
       "| 16nA2 | 0.1169457 | 0.1164594 | 0.0000639 | -1 |   2454 |    254 |   1 |   1 |   1 |   1 |\n",
       "| 16nA3 | 0.7003767 | 0.6932880 | 0.0039033 | -1 | 449639 | 184320 | 512 | 432 | 482 | 400 |\n",
       "| 16nA4 | 0.6771048 | 0.6696726 | 0.0039687 | -1 | 422654 | 181092 | 522 | 444 | 488 | 419 |\n",
       "| 16nA5 | 0.1011897 | 0.1009551 | 0.0000015 | -1 |   2401 |    222 |  NA |  NA |  NA |  NA |\n",
       "| 16nA6 | 0.5432619 | 0.5375346 | 0.0031887 | -1 | 354650 | 144051 | 321 | 266 | 311 | 257 |\n",
       "| 16nA7 | 0.5378295 | 0.5319861 | 0.0030447 | -1 | 348093 | 164068 | 399 | 349 | 384 | 335 |\n",
       "| 16nA8 | 0.0892878 | 0.0891042 | 0.0000072 | -1 |   2586 |    212 |   2 |   1 |   2 |   1 |\n",
       "| 2nA1  | 0.1032909 | 0.1029816 | 0.0000183 | -1 |   1985 |    188 |  NA |  NA |  NA |  NA |\n",
       "| 2nA2  | 0.1311063 | 0.1303647 | 0.0004314 | -1 |  21142 |   8502 |  73 |  70 |  69 |  64 |\n",
       "| 2nA3  | 0.1577292 | 0.1568262 | 0.0000006 | -1 |   2237 |    199 |  NA |  NA |  NA |  NA |\n",
       "| 2nA4  | 0.1071246 | 0.1067223 | 0.0000705 | -1 |   2201 |    253 |   1 |   1 |   1 |   1 |\n",
       "| 2nA5  | 0.1969200 | 0.1944762 | 0.0015129 | -1 |  74984 |  23197 | 146 | 113 | 135 | 103 |\n",
       "| 2nA6  | 0.1436553 | 0.1433478 | 0.0000891 | -1 |   3448 |    283 |  NA |  NA |  NA |  NA |\n",
       "| 2nA7  | 0.0684474 | 0.0683364 | 0.0000024 | -1 |   1693 |    159 |   3 |   2 |   3 |   2 |\n",
       "| 2nA8  | 0.1585038 | 0.1540401 | 0.0033555 | -1 |  51342 |  20866 | 381 | 321 | 356 | 298 |\n",
       "| 4nA1  | 0.0943545 | 0.0941154 | 0.0000177 | -1 |   2362 |    383 |   2 |   2 |   2 |   2 |\n",
       "| 4nA2  | 0.1304181 | 0.1301577 | 0.0000108 | -1 |   2352 |    231 |   2 |   2 |   2 |   2 |\n",
       "| 4nA3  | 0.2298732 | 0.2273217 | 0.0015957 | -1 |  98589 |  31109 | 142 | 111 | 130 | 102 |\n",
       "| 4nA4  | 0.4959204 | 0.4905402 | 0.0030840 | -1 | 224717 |  69245 | 260 | 221 | 247 | 209 |\n",
       "| 4nA5  | 0.1029048 | 0.1026396 | 0.0000255 | -1 |   2464 |    234 |   7 |   4 |   7 |   4 |\n",
       "| 4nA6  | 0.1025673 | 0.1023861 | 0.0000012 | -1 |   2742 |    259 |   1 |   1 |   1 |   1 |\n",
       "| 4nA7  | 0.3244281 | 0.3240933 | 0.0000024 | -1 |   3088 |    213 |   1 |   1 |   1 |   1 |\n",
       "| 4nA8  | 0.3660981 | 0.3618714 | 0.0025242 | -1 | 173405 |  56069 | 199 | 177 | 186 | 164 |\n",
       "| 8nA1  | 0.1311906 | 0.1308564 | 0.0000198 | -1 |   2605 |    250 |   1 |   1 |   1 |   1 |\n",
       "| 8nA2  | 0.1035069 | 0.1031595 | 0.0000174 | -1 |   2281 |    234 |   1 |   1 |   1 |   1 |\n",
       "| 8nA3  | 0.0798825 | 0.0797247 | 0.0000006 | -1 |   2500 |    286 |  NA |  NA |  NA |  NA |\n",
       "| 8nA4  | 0.3374520 | 0.3343263 | 0.0016212 | -1 | 185664 |  86932 | 248 | 221 | 239 | 212 |\n",
       "| 8nA5  | 0.3425886 | 0.3370527 | 0.0036342 | -1 | 199713 |  66713 | 330 | 269 | 315 | 253 |\n",
       "| 8nA6  | 0.0903384 | 0.0901446 | 0.0000009 | -1 |   2789 |    263 |  NA |  NA |  NA |  NA |\n",
       "| 8nA7  | 0.5560356 | 0.5496432 | 0.0034056 | -1 | 352787 | 129172 | 356 | 293 | 340 | 281 |\n",
       "| 8nA8  | 0.4931508 | 0.4875861 | 0.0030840 | -1 | 287061 | 108574 | 325 | 276 | 304 | 259 |\n",
       "\n"
      ],
      "text/plain": [
       "   cellname Rawreads  DNAreads  RNAreads  yperx raw_pairs pairs_dedup UMIs_gene\n",
       "1  16nA1    0.3989055 0.3957042 0.0018330 -1    234464    123576      370      \n",
       "2  16nA2    0.1169457 0.1164594 0.0000639 -1      2454       254        1      \n",
       "3  16nA3    0.7003767 0.6932880 0.0039033 -1    449639    184320      512      \n",
       "4  16nA4    0.6771048 0.6696726 0.0039687 -1    422654    181092      522      \n",
       "5  16nA5    0.1011897 0.1009551 0.0000015 -1      2401       222       NA      \n",
       "6  16nA6    0.5432619 0.5375346 0.0031887 -1    354650    144051      321      \n",
       "7  16nA7    0.5378295 0.5319861 0.0030447 -1    348093    164068      399      \n",
       "8  16nA8    0.0892878 0.0891042 0.0000072 -1      2586       212        2      \n",
       "9  2nA1     0.1032909 0.1029816 0.0000183 -1      1985       188       NA      \n",
       "10 2nA2     0.1311063 0.1303647 0.0004314 -1     21142      8502       73      \n",
       "11 2nA3     0.1577292 0.1568262 0.0000006 -1      2237       199       NA      \n",
       "12 2nA4     0.1071246 0.1067223 0.0000705 -1      2201       253        1      \n",
       "13 2nA5     0.1969200 0.1944762 0.0015129 -1     74984     23197      146      \n",
       "14 2nA6     0.1436553 0.1433478 0.0000891 -1      3448       283       NA      \n",
       "15 2nA7     0.0684474 0.0683364 0.0000024 -1      1693       159        3      \n",
       "16 2nA8     0.1585038 0.1540401 0.0033555 -1     51342     20866      381      \n",
       "17 4nA1     0.0943545 0.0941154 0.0000177 -1      2362       383        2      \n",
       "18 4nA2     0.1304181 0.1301577 0.0000108 -1      2352       231        2      \n",
       "19 4nA3     0.2298732 0.2273217 0.0015957 -1     98589     31109      142      \n",
       "20 4nA4     0.4959204 0.4905402 0.0030840 -1    224717     69245      260      \n",
       "21 4nA5     0.1029048 0.1026396 0.0000255 -1      2464       234        7      \n",
       "22 4nA6     0.1025673 0.1023861 0.0000012 -1      2742       259        1      \n",
       "23 4nA7     0.3244281 0.3240933 0.0000024 -1      3088       213        1      \n",
       "24 4nA8     0.3660981 0.3618714 0.0025242 -1    173405     56069      199      \n",
       "25 8nA1     0.1311906 0.1308564 0.0000198 -1      2605       250        1      \n",
       "26 8nA2     0.1035069 0.1031595 0.0000174 -1      2281       234        1      \n",
       "27 8nA3     0.0798825 0.0797247 0.0000006 -1      2500       286       NA      \n",
       "28 8nA4     0.3374520 0.3343263 0.0016212 -1    185664     86932      248      \n",
       "29 8nA5     0.3425886 0.3370527 0.0036342 -1    199713     66713      330      \n",
       "30 8nA6     0.0903384 0.0901446 0.0000009 -1      2789       263       NA      \n",
       "31 8nA7     0.5560356 0.5496432 0.0034056 -1    352787    129172      356      \n",
       "32 8nA8     0.4931508 0.4875861 0.0030840 -1    287061    108574      325      \n",
       "   genes_gene UMIs_exon genes_exon\n",
       "1  350        356       333       \n",
       "2    1          1         1       \n",
       "3  432        482       400       \n",
       "4  444        488       419       \n",
       "5   NA         NA        NA       \n",
       "6  266        311       257       \n",
       "7  349        384       335       \n",
       "8    1          2         1       \n",
       "9   NA         NA        NA       \n",
       "10  70         69        64       \n",
       "11  NA         NA        NA       \n",
       "12   1          1         1       \n",
       "13 113        135       103       \n",
       "14  NA         NA        NA       \n",
       "15   2          3         2       \n",
       "16 321        356       298       \n",
       "17   2          2         2       \n",
       "18   2          2         2       \n",
       "19 111        130       102       \n",
       "20 221        247       209       \n",
       "21   4          7         4       \n",
       "22   1          1         1       \n",
       "23   1          1         1       \n",
       "24 177        186       164       \n",
       "25   1          1         1       \n",
       "26   1          1         1       \n",
       "27  NA         NA        NA       \n",
       "28 221        239       212       \n",
       "29 269        315       253       \n",
       "30  NA         NA        NA       \n",
       "31 293        340       281       \n",
       "32 276        304       259       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plot <- stat %>% mutate(RNAreadsRatio = RNAreads/(RNAreads+DNAreads),\n",
    "                        pairsPerRead = raw_pairs/DNAreads/1000000000*300,\n",
    "                        pairsValidRatio = pairs_clean3/raw_pairs,\n",
    "                        interPairsRatio = inter_pairs_clean3/pairs_clean3) \n",
    "#plot[is.na(plot)]<-0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   cellname            Rawreads        DNAreads        RNAreads      \n",
       " Length:1           Min.   :1.541   Min.   :1.476   Min.   :0.05569  \n",
       " Class :character   1st Qu.:1.541   1st Qu.:1.476   1st Qu.:0.05569  \n",
       " Mode  :character   Median :1.541   Median :1.476   Median :0.05569  \n",
       "                    Mean   :1.541   Mean   :1.476   Mean   :0.05569  \n",
       "                    3rd Qu.:1.541   3rd Qu.:1.476   3rd Qu.:0.05569  \n",
       "                    Max.   :1.541   Max.   :1.476   Max.   :0.05569  \n",
       "     yperx           raw_pairs       pairs_dedup      pairs_clean1   \n",
       " Min.   :0.02448   Min.   :549679   Min.   :201060   Min.   :200110  \n",
       " 1st Qu.:0.02448   1st Qu.:549679   1st Qu.:201060   1st Qu.:200110  \n",
       " Median :0.02448   Median :549679   Median :201060   Median :200110  \n",
       " Mean   :0.02448   Mean   :549679   Mean   :201060   Mean   :200110  \n",
       " 3rd Qu.:0.02448   3rd Qu.:549679   3rd Qu.:201060   3rd Qu.:200110  \n",
       " Max.   :0.02448   Max.   :549679   Max.   :201060   Max.   :200110  \n",
       "  pairs_clean2     pairs_clean3    inter_pairs_clean3   UMIs_gene   \n",
       " Min.   :162955   Min.   :162875   Min.   :22148      Min.   :6087  \n",
       " 1st Qu.:162955   1st Qu.:162875   1st Qu.:22148      1st Qu.:6087  \n",
       " Median :162955   Median :162875   Median :22148      Median :6087  \n",
       " Mean   :162955   Mean   :162875   Mean   :22148      Mean   :6087  \n",
       " 3rd Qu.:162955   3rd Qu.:162875   3rd Qu.:22148      3rd Qu.:6087  \n",
       " Max.   :162955   Max.   :162875   Max.   :22148      Max.   :6087  \n",
       "   genes_gene     UMIs_exon      genes_exon         1m             200k     \n",
       " Min.   :3245   Min.   :1852   Min.   :1301   Min.   :1.751   Min.   :3.57  \n",
       " 1st Qu.:3245   1st Qu.:1852   1st Qu.:1301   1st Qu.:1.751   1st Qu.:3.57  \n",
       " Median :3245   Median :1852   Median :1301   Median :1.751   Median :3.57  \n",
       " Mean   :3245   Mean   :1852   Mean   :1301   Mean   :1.751   Mean   :3.57  \n",
       " 3rd Qu.:3245   3rd Qu.:1852   3rd Qu.:1301   3rd Qu.:1.751   3rd Qu.:3.57  \n",
       " Max.   :3245   Max.   :1852   Max.   :1301   Max.   :1.751   Max.   :3.57  \n",
       "      20k             50k        RNAreadsRatio      pairsPerRead   \n",
       " Min.   :10.16   Min.   :6.744   Min.   :0.03637   Min.   :0.1118  \n",
       " 1st Qu.:10.16   1st Qu.:6.744   1st Qu.:0.03637   1st Qu.:0.1118  \n",
       " Median :10.16   Median :6.744   Median :0.03637   Median :0.1118  \n",
       " Mean   :10.16   Mean   :6.744   Mean   :0.03637   Mean   :0.1118  \n",
       " 3rd Qu.:10.16   3rd Qu.:6.744   3rd Qu.:0.03637   3rd Qu.:0.1118  \n",
       " Max.   :10.16   Max.   :6.744   Max.   :0.03637   Max.   :0.1118  \n",
       " pairsValidRatio  interPairsRatio\n",
       " Min.   :0.2963   Min.   :0.136  \n",
       " 1st Qu.:0.2963   1st Qu.:0.136  \n",
       " Median :0.2963   Median :0.136  \n",
       " Mean   :0.2963   Mean   :0.136  \n",
       " 3rd Qu.:0.2963   3rd Qu.:0.136  \n",
       " Max.   :0.2963   Max.   :0.136  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.matrix.max.cols=50, repr.matrix.max.rows=100)\n",
    "plot %>% summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A rowwise_df: 1 × 23</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cellname</th><th scope=col>Rawreads</th><th scope=col>DNAreads</th><th scope=col>RNAreads</th><th scope=col>yperx</th><th scope=col>raw_pairs</th><th scope=col>pairs_dedup</th><th scope=col>pairs_clean1</th><th scope=col>pairs_clean2</th><th scope=col>pairs_clean3</th><th scope=col>inter_pairs_clean3</th><th scope=col>UMIs_gene</th><th scope=col>genes_gene</th><th scope=col>UMIs_exon</th><th scope=col>genes_exon</th><th scope=col>1m</th><th scope=col>200k</th><th scope=col>20k</th><th scope=col>50k</th><th scope=col>RNAreadsRatio</th><th scope=col>pairsPerRead</th><th scope=col>pairsValidRatio</th><th scope=col>interPairsRatio</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>OrgfE951001</td><td>1.540634</td><td>1.47555</td><td>0.0556875</td><td>0.02448</td><td>549679</td><td>201060</td><td>200110</td><td>162955</td><td>162875</td><td>22148</td><td>6087</td><td>3245</td><td>1852</td><td>1301</td><td>1.751261</td><td>3.570228</td><td>10.15922</td><td>6.744294</td><td>0.03636765</td><td>0.1117575</td><td>0.2963093</td><td>0.1359816</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A rowwise\\_df: 1 × 23\n",
       "\\begin{tabular}{lllllllllllllllllllllll}\n",
       " cellname & Rawreads & DNAreads & RNAreads & yperx & raw\\_pairs & pairs\\_dedup & pairs\\_clean1 & pairs\\_clean2 & pairs\\_clean3 & inter\\_pairs\\_clean3 & UMIs\\_gene & genes\\_gene & UMIs\\_exon & genes\\_exon & 1m & 200k & 20k & 50k & RNAreadsRatio & pairsPerRead & pairsValidRatio & interPairsRatio\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t OrgfE951001 & 1.540634 & 1.47555 & 0.0556875 & 0.02448 & 549679 & 201060 & 200110 & 162955 & 162875 & 22148 & 6087 & 3245 & 1852 & 1301 & 1.751261 & 3.570228 & 10.15922 & 6.744294 & 0.03636765 & 0.1117575 & 0.2963093 & 0.1359816\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A rowwise_df: 1 × 23\n",
       "\n",
       "| cellname &lt;chr&gt; | Rawreads &lt;dbl&gt; | DNAreads &lt;dbl&gt; | RNAreads &lt;dbl&gt; | yperx &lt;dbl&gt; | raw_pairs &lt;dbl&gt; | pairs_dedup &lt;dbl&gt; | pairs_clean1 &lt;dbl&gt; | pairs_clean2 &lt;dbl&gt; | pairs_clean3 &lt;dbl&gt; | inter_pairs_clean3 &lt;dbl&gt; | UMIs_gene &lt;dbl&gt; | genes_gene &lt;dbl&gt; | UMIs_exon &lt;dbl&gt; | genes_exon &lt;dbl&gt; | 1m &lt;dbl&gt; | 200k &lt;dbl&gt; | 20k &lt;dbl&gt; | 50k &lt;dbl&gt; | RNAreadsRatio &lt;dbl&gt; | pairsPerRead &lt;dbl&gt; | pairsValidRatio &lt;dbl&gt; | interPairsRatio &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| OrgfE951001 | 1.540634 | 1.47555 | 0.0556875 | 0.02448 | 549679 | 201060 | 200110 | 162955 | 162875 | 22148 | 6087 | 3245 | 1852 | 1301 | 1.751261 | 3.570228 | 10.15922 | 6.744294 | 0.03636765 | 0.1117575 | 0.2963093 | 0.1359816 |\n",
       "\n"
      ],
      "text/plain": [
       "  cellname    Rawreads DNAreads RNAreads  yperx   raw_pairs pairs_dedup\n",
       "1 OrgfE951001 1.540634 1.47555  0.0556875 0.02448 549679    201060     \n",
       "  pairs_clean1 pairs_clean2 pairs_clean3 inter_pairs_clean3 UMIs_gene\n",
       "1 200110       162955       162875       22148              6087     \n",
       "  genes_gene UMIs_exon genes_exon 1m       200k     20k      50k     \n",
       "1 3245       1852      1301       1.751261 3.570228 10.15922 6.744294\n",
       "  RNAreadsRatio pairsPerRead pairsValidRatio interPairsRatio\n",
       "1 0.03636765    0.1117575    0.2963093       0.1359816      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.0.3",
   "language": "R",
   "name": "ir403"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
